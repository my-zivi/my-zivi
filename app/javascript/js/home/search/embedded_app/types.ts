/* eslint-disable camelcase */

export interface JobPostingSearchHit {
  objectID: string;
  title: string;
  icon_url: string;
  publication_date: string;
  plain_description: string;
  description: string;
  brief_description: string;
  required_skills?: string;
  preferred_skills?: string;
  canton: string;
  identification_number: string;
  relevancy: Relevancy;
  contact_information: string;
  minimum_service_months: number;
  language: string;
  organization_display_name: string;
  category_display_name: string;
  sub_category_display_name?: string;
  link: string;
  featured_as_new: boolean;
  priority_program: boolean;
}

export enum Relevancy {
  Normal = 1,
  Urgent,
  Featured
}
