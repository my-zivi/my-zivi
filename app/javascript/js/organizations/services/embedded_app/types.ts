export interface Service {
  beginning: string;
  ending: string;
  civilServant: {
    fullName: string;
  };
}
