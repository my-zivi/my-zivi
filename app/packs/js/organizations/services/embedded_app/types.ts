export interface Service {
  beginning: string;
  ending: string;
  definitive: boolean;
  civilServant: {
    fullName: string;
  };
}
