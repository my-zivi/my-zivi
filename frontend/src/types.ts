import { FormikBag } from 'formik';

export interface Holiday {
  id?: number;
  beginning: string;
  ending: string;
  holiday_type: 'public_holiday' | 'company_holiday';
  description: string;
}

export enum PaymentState {
  payment_in_progress = 'payment_in_progress',
  paid = 'paid',
}

export interface Payment {
  id?: number;
  total: number;
  state: PaymentState;
  payment_timestamp: number;
  expense_sheets: PaymentExpenseSheet[];
}

export interface PaymentExpenseSheet {
  id?: number;
  total: number;
  user: {
    id: number;
    bank_iban: string;
    full_name: string;
    zdp: number;
  };
}

export interface ExpenseSheet {
  id: number;
  bank_account_number: string;
  beginning: Date;
  clothing_expenses: number;
  clothing_expenses_comment: null;
  company_holiday_comment: null;
  driving_expenses: number;
  driving_expenses_comment: null;
  duration: number;
  ending: Date;
  extraordinary_expenses: number;
  extraordinary_expenses_comment: null;
  paid_company_holiday_days: number;
  paid_vacation_comment: null;
  paid_vacation_days: number;
  payment_timestamp: null;
  sick_comment: null;
  sick_days: number;
  state: ExpenseSheetState;
  total: number;
  unpaid_company_holiday_days: number;
  unpaid_vacation_comment: null;
  unpaid_vacation_days: number;
  user_id: number;
  work_days: number;
  workfree_days: number;
  service_id: number;
}

export interface ExpenseSheetHints {
  suggestions: {
    work_days: number;
    workfree_days: number;
    paid_company_holiday_days: number;
    unpaid_company_holiday_days: number;
    clothing_expenses: number;
  };
  remaining_days: {
    sick_days: number;
    paid_vacation_days: number;
  };
}

export enum ExpenseSheetState {
  open = 'open',
  ready_for_payment = 'ready_for_payment',
  payment_in_progress = 'payment_in_progress',
  paid = 'paid',
}

export interface ShortExpenseSheetListing {
  id: number;
  ending: string;
  beginning: string;
  state: ExpenseSheetState;
  duration: number;
}

export interface ExpenseSheetListing extends ShortExpenseSheetListing {
  total: number;
  user: {
    id: number;
    address: string;
    bank_iban: string;
    city: string;
    full_name: string;
    zdp: number;
    zip: number;
  };
}

export interface DailyExpense {
  breakfast: number;
  lunch: number;
  dinner: number;
}

export interface ServiceSpecification {
  id: number;
  identification_number?: number;
  name: string;
  short_name: string;
  work_clothing_expenses: number;
  work_days_expenses: DailyExpense;
  paid_vacation_expenses: DailyExpense;
  first_day_expenses: DailyExpense;
  last_day_expenses: DailyExpense;
  accommodation_expenses: number;
  pocket_money: number;
  active: boolean;
}

export interface User {
  id: number;
  active: boolean;
  address: string;
  bank_iban: string;
  birthday: string;
  chainsaw_workshop: boolean;
  city: string;
  driving_licence_b: boolean;
  driving_licence_be: boolean;
  email: string;
  ending: null | string;
  first_name: string;
  health_insurance: string;
  hometown: string;
  internal_note: string;
  last_name: string;
  phone: string;
  regional_center_id: number;
  expense_sheets: ShortExpenseSheetListing[];
  role: 'admin' | 'civil_servant';
  services: Service[];
  beginning: null | string;
  work_experience: null | string;
  zdp: number;
  zip: number | null;
}

export interface RegionalCenter {
  name: string;
  address: string;
  short_name: string;
  id: number;
}

export interface UserFilter {
  zdp: string;
  name: string;
  beginning: string;
  ending: string;
  active: boolean;
  role: string;
}

export interface Service {
  id?: number;
  beginning: Date | null;
  service_days: number;
  confirmation_date: null | string;
  eligible_paid_vacation_days: number;
  ending: Date | null;
  first_swo_service: boolean;
  long_service: boolean;
  service_type: string | null;
  probation_period: boolean;
  service_specification_id: number;
  service_specification: {
    identification_number: number;
    name: string | undefined;
    short_name: string | undefined;
  };
  user_id: number;
}

export interface ServiceCollection {
  id?: number;
  beginning: string | null;
  ending: string | null;
  confirmation_date: string | null;
  service_specification: {
    identification_number: number;
    name: string | undefined;
    short_name: string | undefined;
  };
  user: {
    id: number;
    first_name: string | null;
    last_name: string | null;
    zdp: number;
  };
}

export interface UserFeedback {
  id?: number;
}

export interface UserQuestionWithAnswers {
  id?: number;
  answers: UserQuestionAnswers;
  custom_info: string;
  opt1: string;
  opt2: string;
  question: string;
  type: number;
}

export interface UserQuestionAnswers {
  1: number;
  2: number;
  3: number;
  4: number;
  5: number;
  6: number;
}

export interface Listing {
  id?: number;
  archived?: boolean;
}

export interface Column<T> {
  id: string;
  numeric?: boolean;
  label: string;
  format?: (t: T) => React.ReactNode;
  span?: {
    col?: number;
    row?: number;
  };
}

export type ActionButtonAction = (() => void) | string;

// tslint:disable-next-line:no-any ; really don't care for that type, and it comes from deep inside Formik
export type HandleFormikSubmit<Values> = (values: Values, formikBag: FormikBag<any, Values>) => void;

// If we'd type thoroughly we'd need to create a type for each models representation in a form / yup validation schema
// tslint:disable-next-line:no-any
export type FormValues = any;
