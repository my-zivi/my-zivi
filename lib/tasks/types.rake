# frozen_string_literal: true

def js_type(indentation, value)
  case value
  when String
    'string'
  when Array
    'string[]'
  else
    to_js_object(value, indentation + 2)
  end
end

def to_js_object(hash, indentation = 2)
  content = hash.map do |key, value|
    indent = ' ' * indentation

    "#{indent}#{key}: #{js_type(indentation, value)};"
  end

  base_indent = ' ' * (indentation - 2)

  "{\n#{content.join("\n")}\n#{base_indent}}"
end

namespace :types do
  desc 'Generates Typescript translation types'
  task generate: :environment do
    content = to_js_object(I18n.t('search').deep_transform_keys { |key| key.to_s.camelize(:lower) })

    template = <<~JS
      // Generated file using `rake types:generate`

      declare interface SearchTranslations #{content[..-1]}
    JS

    File.write(Rails.root.join('app/javascript/js/shared/SearchTranslations.d.ts'), template)
  end
end
