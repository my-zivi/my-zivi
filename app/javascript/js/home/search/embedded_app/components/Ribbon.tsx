/* eslint-disable camelcase */
import { Relevancy, JobPostingSearchHit } from 'js/home/search/embedded_app/types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'preact/compat';

const Ribbon: React.FunctionComponent<{hit: JobPostingSearchHit}> = ({ hit: { relevancy, featured_as_new } }) => {
  if (relevancy === Relevancy.Urgent) {
    return (
      <div className="featured-ribbon red">
        <FontAwesomeIcon icon={['fas', 'exclamation']} className="mr-2" />
        {MyZivi.translations.search.ribbons.urgent}
      </div>
    );
  }

  if (relevancy === Relevancy.Featured) {
    return (
      <div className="featured-ribbon yellow">
        <FontAwesomeIcon icon={['fas', 'star']} className={'mr-1'} />
        {MyZivi.translations.search.ribbons.featured}
      </div>
    );
  }

  if (featured_as_new) {
    return <div className="featured-ribbon blue">{MyZivi.translations.search.ribbons.new}</div>;
  }

  return <></>;
};

export default Ribbon;
