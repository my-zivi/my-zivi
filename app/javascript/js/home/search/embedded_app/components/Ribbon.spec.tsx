import { shallow } from 'enzyme';
import React from 'preact/compat';
import Ribbon from './Ribbon';
import JobPostingFactory from '../../../../tests/factories/JobPostingFactory';
import { Relevancy } from '../types';

describe('Ribbon', () => {
  it('renders correctly', () => {
    expect(
      shallow(<Ribbon hit={JobPostingFactory.build()} />),
    ).toMatchSnapshot();
  });

  describe('when job posting is new', () => {
    expect(
      shallow(<Ribbon hit={JobPostingFactory.build({ featured_as_new: true })} />),
    ).toMatchSnapshot();
  });

  describe('when job posting is urgent', () => {
    expect(
      shallow(<Ribbon hit={JobPostingFactory.build({ relevancy: Relevancy.Urgent })} />),
    ).toMatchSnapshot();
  });

  describe('when job posting is featured', () => {
    expect(
      shallow(<Ribbon hit={JobPostingFactory.build({ relevancy: Relevancy.Featured })} />),
    ).toMatchSnapshot();
  });
});
