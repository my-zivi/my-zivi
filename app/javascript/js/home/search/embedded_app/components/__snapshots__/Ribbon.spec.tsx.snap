// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="featured-ribbon blue"
>
  Neu
</div>
`;

exports[` 2`] = `
<div
  className="featured-ribbon red"
>
  <FontAwesomeIcon
    className="mr-2"
    icon={
      Array [
        "fas",
        "exclamation",
      ]
    }
  />
  Wichtig
</div>
`;

exports[` 3`] = `
<div
  className="featured-ribbon yellow"
>
  <FontAwesomeIcon
    className="mr-1"
    icon={
      Array [
        "fas",
        "star",
      ]
    }
  />
  Feat.
</div>
`;

exports[`Ribbon renders correctly 1`] = `null`;
